### PR TITLE
Various kubenet fixes (panics and bugs and cidrs, oh my)

### PR DESCRIPTION
**Pull request by euank**
This PR fixes the following issues:
1. Corrects an inverse error-check that prevented `shaper.Reset` from ever being called with a correct ip address
2. Fix an issue where `parseCIDR` would fail after a kubelet restart due to an IP being stored instead of a CIDR being stored in the cache.
3. Fix an issue where kubenet could panic in TearDownPod if it was called before SetUpPod (e.g. after a kubelet restart).. because of bug number 1, this didn't happen except in rare situations (see 2 for why such a rare situation might happen)

This adds a test, but more would definitely be useful.
The commits are also granular enough I could split this up more if desired.

I'm also not super-familiar with this code, so review and feedback would be welcome.

Testing done:

```
$ cat examples/egress/egress.yml
 apiVersion: v1
kind: Pod
metadata:
  labels:
    name: egress
  name: egress-output
  annotations: {"kubernetes.io/ingress-bandwidth": "300k"}
spec:
  restartPolicy: Never
  containers:
    - name: egress
      image: busybox
      command: ["sh", "-c", "sleep 60"]
$ cat kubelet.log
...
Running: tc filter add dev cbr0 protocol ip parent 1:0 prio 1 u32 match ip dst 10.0.0.5/32 flowid 1:1
# setup
...
Running: tc filter del dev cbr0 parent 1:proto ip prio 1 handle 800::800 u32
# teardown
```

I also did various other bits of manual testing and logging to hunt down the panic and other issues, but don't have anything to paste for that 

cc @dcbw @kubernetes/sig-network

### CODE REVIEW

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -317,10 +317,10 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
 	if err != nil {
 		return err
 	}
-	if res.IP4 == nil || res.IP4.IP.String() == "" {
+	if res.IP4 == nil || res.IP4.IP.IP.String() == "" {
```
Comment (euank) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Don't both net.IP and net.IPAddr's `String` print `"<nil>"` (even for technically non-nil structs) for invalid stuff, not `""`? Is this right?

Comment (dcbw) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: You're right, it returns either <nil> or ? for invalid addresses.  Looks like the correct check should actually be len(res.IP4.IP.IP) == IPv4len.

**Comment by dcbw** [comment]
LGTM after the IP String() fixup is done, and now I've got some other code to fix up elsewhere for the same reason :)

**Comment by euank** [comment]
Updated to a `len` check as suggested and rebased.

**Comment by thockin** [comment]
either assigneee can approve

**Comment by dcbw** [comment]
Looks like the e2e test did uncover an issue:

May 27 11:25:33.744: INFO: At {2016-05-27 11:12:48 -0700 PDT} - event for e2e-image-puller-e2e-gce-master-0-minion-group-kxcy: {kubelet e2e-gce-master-0-minion-group-kxcy} FailedSync: Error syncing pod, skipping: failed to "SetupNetwork" for "e2e-image-puller-e2e-gce-master-0-minion-group-kxcy_kube-system" with SetupNetworkError: "Failed to setup network for pod \"e2e-image-puller-e2e-gce-master-0-minion-group-kxcy_kube-system(2b2e55c843c1c2c9c7544e84962e4373)\" using network plugins \"kubenet\": CNI plugin reported no IPv4 address for container {docker 057ab31c3f248c7fcc8ef55b80dc9a6ade0fab9eb3df8b2d0bb55c27e6af0b61}.; Skipping pod"

**Comment by euank** [comment]
I guess correcting the ip check uncovered another bad assumption.. I'll look into it further.

**Comment by dcbw** [comment]
@euank yeah, I'm looking too.  Looks like we probably want to use something like:
```
	if res.IP4 == nil {
		return fmt.Errorf("CNI plugin reported no IPv4 address for container %v.", id)
	}
	if ip := res.IP4.IP.IP.To4(); ip == nil {
		return fmt.Errorf("CNI plugin reported invalid IPv4 address for container %v.", id)
	}
```

**Comment by dcbw** [comment]
You'll probably also need a rebase once #26208 lands too.

**Comment by euank** [comment]
TIL that `len(IP)` can be 16 for a valid ipv4 address in Go. Cool.

Anyways, updated to something quite similar to what you suggested. I'm glad the test caught that :smile:.

**Comment by freehan** [comment]
Great caught! I always ignore the shaper piece... 

Just curious, why bug number 1 could cause kubelet restart?

**Comment by dcbw** [comment]
lgtm assuming the tests pass...

**Comment by dchen1107** [comment]
Marked this for 1.3 milestone and bump the priority to p1.

**Comment by freehan** [comment]
Oh, I misread the description. Please ignore my question

**Comment by freehan** [comment]
LGTM
pending tests

**Pull request merged**